File: src/lib/migration.ts
/**
 * One-time migration from localStorage to SQLite.
 * Runs on first boot after update; idempotent via a localStorage flag.
 */

import { db } from "./db";
import type {
  GatewayRow,
  SettingsRow,
  DeviceIdentityRow,
  DeviceAuthRow,
  NpcThemeRow,
  BgPathEntry,
  GatewaySessionEntry,
  GatewayThemeEntry,
  GatewayDeviceIdEntry,
  MigrationPayload,
} from "./db";

const MIGRATION_FLAG = "openclaw.db_migrated";

export async function migrateIfNeeded(): Promise<void> {
  if (localStorage.getItem(MIGRATION_FLAG)) return;

  console.log("[Migration] Starting localStorage → SQLite migration...");

  try {
    const payload = collectLegacyData();

    // Only migrate if there's actually data to migrate
    const hasData =
      payload.gateways?.length ||
      payload.settings ||
      payload.deviceIdentity ||
      payload.deviceAuthEntries?.length ||
      payload.customThemes?.length;

    if (hasData) {
      await db.migrate(payload);
      console.log("[Migration] Migration completed successfully");
    } else {
      console.log("[Migration] No legacy data found, skipping");
    }

    localStorage.setItem(MIGRATION_FLAG, "1");
  } catch (e) {
    console.error("[Migration] Migration failed:", e);
    // Don't set the flag — retry on next boot
  }
}

function collectLegacyData(): MigrationPayload {
  const payload: MigrationPayload = {};

  // ── Gateways ──────────────────────────────────────────
  try {
    const raw = localStorage.getItem("openclaw.gateways");
    if (raw) {
      const data = JSON.parse(raw) as {
        gateways?: Array<{
          id: string;
          name: string;
          url: string;
          authMethod?: string;
          token?: string;
          password?: string;
          deviceToken?: string;
        }>;
        activeId?: string;
      };

      if (data.gateways?.length) {
        payload.gateways = data.gateways.map((g) => ({
          id: g.id,
          name: g.name,
          url: g.url,
          authMethod: g.authMethod ?? "token",
          token: g.token ?? null,
          password: g.password ?? null,
          deviceToken: g.deviceToken ?? null,
          activeSessionKey: "main",
          activeNpcThemeId: "default",
          deviceId: null,
          sortOrder: 0,
        }));
      }

      if (data.activeId) {
        payload.activeGatewayId = data.activeId;
      }
    }
  } catch (e) {
    console.warn("[Migration] Failed to read gateways:", e);
  }

  // ── Per-gateway session keys ──────────────────────────
  try {
    const sessions: GatewaySessionEntry[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith("openclaw.session.")) {
        const gatewayId = key.slice("openclaw.session.".length);
        const sessionKey = localStorage.getItem(key);
        if (gatewayId && sessionKey) {
          sessions.push({ gatewayId, sessionKey });
        }
      }
    }
    if (sessions.length) payload.gatewaySessions = sessions;
  } catch (e) {
    console.warn("[Migration] Failed to read sessions:", e);
  }

  // ── Per-gateway active NPC theme IDs ──────────────────
  try {
    const themeIds: GatewayThemeEntry[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith("openclaw.npcActiveTheme.")) {
        const gatewayId = key.slice("openclaw.npcActiveTheme.".length);
        const themeId = localStorage.getItem(key);
        if (gatewayId && themeId) {
          themeIds.push({ gatewayId, themeId });
        }
      }
    }
    if (themeIds.length) payload.gatewayThemeIds = themeIds;
  } catch (e) {
    console.warn("[Migration] Failed to read theme IDs:", e);
  }

  // ── Per-gateway device IDs ────────────────────────────
  try {
    const deviceIds: GatewayDeviceIdEntry[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith("openclaw.deviceId.")) {
        const gatewayId = key.slice("openclaw.deviceId.".length);
        const deviceId = localStorage.getItem(key);
        if (gatewayId && deviceId) {
          deviceIds.push({ gatewayId, deviceId });
        }
      }
    }
    if (deviceIds.length) payload.gatewayDeviceIds = deviceIds;
  } catch (e) {
    console.warn("[Migration] Failed to read device IDs:", e);
  }

  // ── Settings ──────────────────────────────────────────
  try {
    const raw = localStorage.getItem("openclaw.settings");
    if (raw) {
      const parsed = JSON.parse(raw) as {
        autoUpdate?: boolean;
        launchOnStartup?: boolean;
        minimizeToTray?: boolean;
        apiKeys?: Record<string, string>;
      };
      payload.settings = {
        autoUpdate: parsed.autoUpdate ?? true,
        launchOnStartup: parsed.launchOnStartup ?? false,
        minimizeToTray: parsed.minimizeToTray ?? true,
        apiKeys: JSON.stringify(parsed.apiKeys ?? {}),
      };
    }
  } catch (e) {
    console.warn("[Migration] Failed to read settings:", e);
  }

  // ── Device Identity ───────────────────────────────────
  try {
    const raw = localStorage.getItem("openclaw-device-identity-v1");
    if (raw) {
      const parsed = JSON.parse(raw) as {
        version: number;
        deviceId: string;
        publicKey: string;
        privateKey: string;
      };
      if (parsed.version === 1 && parsed.deviceId && parsed.publicKey && parsed.privateKey) {
        payload.deviceIdentity = {
          deviceId: parsed.deviceId,
          publicKey: parsed.publicKey,
          privateKey: parsed.privateKey,
        };
      }
    }
  } catch (e) {
    console.warn("[Migration] Failed to read device identity:", e);
  }

  // ── Device Auth ───────────────────────────────────────
  try {
    const raw = localStorage.getItem("openclaw.device.auth.v1");
    if (raw) {
      const parsed = JSON.parse(raw) as {
        version: number;
        deviceId: string;
        tokens: Record<string, { token: string; role: string; scopes: string[] }>;
      };
      if (parsed.version === 1 && parsed.tokens) {
        const entries: DeviceAuthRow[] = [];
        for (const [role, entry] of Object.entries(parsed.tokens)) {
          entries.push({
            deviceId: parsed.deviceId,
            role,
            token: entry.token,
            scopes: JSON.stringify(entry.scopes ?? []),
          });
        }
        if (entries.length) payload.deviceAuthEntries = entries;
      }
    }
  } catch (e) {
    console.warn("[Migration] Failed to read device auth:", e);
  }

  // ── Custom NPC Themes ─────────────────────────────────
  try {
    const raw = localStorage.getItem("openclaw.npcThemes");
    if (raw) {
      const themes = JSON.parse(raw) as Array<{
        id: string;
        name: string;
        description?: string;
        [key: string]: unknown;
      }>;
      payload.customThemes = themes.map((t) => ({
        id: t.id,
        name: t.name,
        description: t.description ?? null,
        data: JSON.stringify(t),
      }));
    }
  } catch (e) {
    console.warn("[Migration] Failed to read custom themes:", e);
  }

  // ── Background Paths ──────────────────────────────────
  try {
    const bgPaths: BgPathEntry[] = [];
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (key?.startsWith("openclaw.npcBgPath.")) {
        const themeId = key.slice("openclaw.npcBgPath.".length);
        const filePath = localStorage.getItem(key);
        if (themeId && filePath) {
          bgPaths.push({ themeId, filePath });
        }
      }
    }
    if (bgPaths.length) payload.bgPaths = bgPaths;
  } catch (e) {
    console.warn("[Migration] Failed to read bg paths:", e);
  }

  return payload;
}

File: src-tauri/src/database/migration.rs
use super::models::MigrationPayload;
use super::DbState;
use tauri::State;

#[tauri::command]
pub fn db_migrate_from_json(db: State<'_, DbState>, data: MigrationPayload) -> Result<(), String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;

    // Run everything in a transaction for atomicity
    conn.execute("BEGIN TRANSACTION", [])
        .map_err(|e| e.to_string())?;

    let result = (|| -> Result<(), String> {
        // Migrate gateways
        if let Some(gateways) = &data.gateways {
            for gw in gateways {
                conn.execute(
                    "INSERT OR IGNORE INTO gateways (id, name, url, auth_method, token, password, device_token)
                     VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
                    rusqlite::params![
                        gw.id, gw.name, gw.url, gw.auth_method,
                        gw.token, gw.password, gw.device_token,
                    ],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate active gateway ID
        if let Some(active_id) = &data.active_gateway_id {
            conn.execute(
                "INSERT OR REPLACE INTO app_state (key, value) VALUES ('active_gateway_id', ?1)",
                [active_id],
            )
            .map_err(|e| e.to_string())?;
        }

        // Migrate per-gateway session keys
        if let Some(sessions) = &data.gateway_sessions {
            for entry in sessions {
                conn.execute(
                    "UPDATE gateways SET active_session_key = ?1 WHERE id = ?2",
                    rusqlite::params![entry.session_key, entry.gateway_id],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate per-gateway active NPC theme IDs
        if let Some(theme_ids) = &data.gateway_theme_ids {
            for entry in theme_ids {
                conn.execute(
                    "UPDATE gateways SET active_npc_theme_id = ?1 WHERE id = ?2",
                    rusqlite::params![entry.theme_id, entry.gateway_id],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate per-gateway device IDs
        if let Some(device_ids) = &data.gateway_device_ids {
            for entry in device_ids {
                conn.execute(
                    "UPDATE gateways SET device_id = ?1 WHERE id = ?2",
                    rusqlite::params![entry.device_id, entry.gateway_id],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate settings
        if let Some(s) = &data.settings {
            conn.execute(
                "INSERT OR REPLACE INTO settings (id, auto_update, launch_on_startup, minimize_to_tray, api_keys)
                 VALUES (1, ?1, ?2, ?3, ?4)",
                rusqlite::params![
                    s.auto_update as i32,
                    s.launch_on_startup as i32,
                    s.minimize_to_tray as i32,
                    s.api_keys,
                ],
            )
            .map_err(|e| e.to_string())?;
        }

        // Migrate device identity
        if let Some(identity) = &data.device_identity {
            conn.execute(
                "INSERT OR REPLACE INTO device_identity (id, device_id, public_key, private_key)
                 VALUES (1, ?1, ?2, ?3)",
                rusqlite::params![identity.device_id, identity.public_key, identity.private_key],
            )
            .map_err(|e| e.to_string())?;
        }

        // Migrate device auth entries
        if let Some(entries) = &data.device_auth_entries {
            for entry in entries {
                conn.execute(
                    "INSERT OR REPLACE INTO device_auth (device_id, role, token, scopes)
                     VALUES (?1, ?2, ?3, ?4)",
                    rusqlite::params![entry.device_id, entry.role, entry.token, entry.scopes],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate custom NPC themes
        if let Some(themes) = &data.custom_themes {
            for theme in themes {
                conn.execute(
                    "INSERT OR IGNORE INTO npc_themes (id, name, description, data)
                     VALUES (?1, ?2, ?3, ?4)",
                    rusqlite::params![theme.id, theme.name, theme.description, theme.data],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        // Migrate background paths
        if let Some(paths) = &data.bg_paths {
            for entry in paths {
                conn.execute(
                    "INSERT OR IGNORE INTO npc_bg_paths (theme_id, file_path)
                     VALUES (?1, ?2)",
                    rusqlite::params![entry.theme_id, entry.file_path],
                )
                .map_err(|e| e.to_string())?;
            }
        }

        Ok(())
    })();

    match result {
        Ok(()) => {
            conn.execute("COMMIT", []).map_err(|e| e.to_string())?;
            Ok(())
        }
        Err(e) => {
            let _ = conn.execute("ROLLBACK", []);
            Err(e)
        }
    }
}
